File: 264. Ugly Number II/solution.py
class Solution:
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 1:
            return 1
        
        uglies = [1]
        p2 = 0
        p3 = 0
        p5 = 0
        
        while len(uglies) != n:
            ugly = min(2*uglies[p2], 3*uglies[p3], 5*uglies[p5])
            if 2*uglies[p2] == ugly:
                p2 += 1
            elif 3*uglies[p3] == ugly:
                p3 += 1
            elif 5*uglies[p5] == ugly:
                p5 += 1
                
            if ugly > uglies[-1]:
                uglies.append(ugly)
                
        return uglies[-1]

File: 264. Ugly Number II/solution.java
class Solution {
    public int nthUglyNumber(int n) {
        if (n == 0) {
            return 0;
        }
        
        if (n == 1) {
            return 1;
        }
        
        int[] ugly = new int[n];
        int x = 0;
        int y = 0;
        int z = 0;
        int i = 0;
        ugly[i] = 1;
        i++;
        while (i < n) {
            ugly[i] = Math.min(ugly[x]*2, Math.min(ugly[y]*3, ugly[z]*5));
            if (ugly[i] == ugly[x]*2) {
                x++;
            } else if (ugly[i] == ugly[y]*3) {
                y++;
            } else {
                z++;
            } 
            
            if (ugly[i] > ugly[i-1]) {
                i++;
            }
        }
        return ugly[i-1];
    }
}
